text/0000-build-std/2-history.md

```diff
@@ -0,0 +1,543 @@
+
+Since around 2020, activity in the [wg-cargo-std-aware] repository largely
+trailed off and there have not been any significant developments related to
+build-std since.
+
+### Implementation summary
+[implementation-summary]: #implementation-summary
+
+*An exhaustive review of implementation-related issues, pull requests and
+discussions can be found in
+[the relevant section of the literature review][appendix-ii-impl].*
+
+There has been an unstable and experimental implementation of build-std in Cargo
+since August 2019 ([wg-cargo-std-aware#10]/[cargo#7216]).
+
+[cargo#7216] added the [`-Zbuild-std`][build-std] flag to Cargo. `-Zbuild-std`
+re-builds the standard library crates which rustc then uses instead of the
+pre-built standard library from the sysroot.
+
+Originally, `-Zbuild-std` always build `std` by default. Since the addition of
+the `std` field to target metadata in [rust#122305], Cargo only builds `std` by
+default if `metadata.std` is true.
+
+`test` is also built if `std` is being built and tests are being run with the
+default harness.
+
+Optionally, users can provide the list of crates to be built, though this was
+intended as an escape hatch to work around bugs - the arguments to the flag are
+unstable since the names of crates comprising the standard library are not
+stable.
+
+Cargo has a hardcoded list of what dependencies need to be added for a given
+user-requested crate (i.e. `std` implies building `core`, `alloc`,
+`compiler_builtins`, etc.). It is common for users to manually specify the
+`panic_abort` crate.
+
+Originally, `-Zbuild-std` required that `--target` be provided
+([wg-cargo-std-aware#25]) to force Cargo to use different sysroots for the host
+and target , but this restriction was later resolved ([cargo#14317]).
+
+A second flag, [`-Zbuild-std-features`][build-std-features], was added in
+[cargo#8490] and allows overriding the default Cargo features of the standard
+library. Like the arguments to `-Zbuild-std`, this values accepted by this flag
+are inherently unstable as the library team has not committed to any of the
+standard library's Cargo features being stable. Features are enabled on the
+`sysroot` crate and propagate down through the crate graph of the standard
+library (e.g. `compiler-builtins-mem` is a feature in `sysroot`, `std`,
+`alloc`, and `core` until `compiler_builtins`).
+
+build-std gets the source of the standard library from the `rust-src` rustup
+component. This does not happen automatically and the user must ensure the
+component has been downloaded themselves. Only the standard library crates from
+the [rust-lang/rust] repository are included in the `rust-src` dependency (i.e.
+none of the crates.io dependencies).
+
+When `-Zbuild-std` has been passed, Cargo creates a second workspace for the
+standard library based on the `Cargo.{toml,lock}` from the `rust-src` component.
+Originally this was a virtual workspace, prior to the standard library having a
```

Comment:
I *think* this would be more technically correct
```suggestion
Originally this was an ephemeral workspace, prior to the standard library having a
```

Reply:
I don't believe the virtual workspace was ephemeral, at least at the point it was removed. See [this comment](https://github.com/rust-lang/cargo/pull/14358/files#diff-4729807f734690fc5a8a12545fa6bc9ccf1b9a87c91969d1af5f0803f714a769L128-L130) and the workspace constructor that was used before being removed

Reply:
*sigh* so many different but slightly different ways of doing things.

For the intent of what is being said, "an in-memory workspace" would probably be best than.  `virtual` has specific meaning in cargo and I think this is using it in a way different than cargo (it was both an in-memory and virtual workspace but for the context here, the in-memory part is more relevant).

Reply:
Fixed in a99cd20

Reply:
Thanks!